configuration/configurator/SchemaKey.cpp

```diff
@@ -0,0 +1,222 @@
+
+#include "jptree.hpp"
+#include "SchemaKey.hpp"
+#include "SchemaSelector.hpp"
+#include "SchemaField.hpp"
+#include "SchemaAnnotation.hpp"
+#include "SchemaCommon.hpp"
+#include "ConfigSchemaHelper.hpp"
+#include "SchemaMapManager.hpp"
+#include "SchemaAttributes.hpp"
+
+using namespace CONFIGURATOR;
+
+#define IPropertyTree ::IPropertyTree
+
+CKey* CKey::load(CXSDNodeBase* pParentNode, const IPropertyTree *pSchemaRoot, const char* xpath)
+{
+    assert(pSchemaRoot != NULL);
+    assert(pParentNode != NULL);
+    assert(pParentNode->getNodeType() == XSD_KEY_ARRAY);
+
+    if (pSchemaRoot == NULL || pParentNode == NULL)
+    {
+        // TODO: Throw Exception
+        return NULL;
+    }
+
+    CKey *pKey = new CKey(pParentNode);
+
+    if (xpath != NULL && *xpath != 0)
+    {
+        IPropertyTree* pTree = pSchemaRoot->queryPropTree(xpath);
+
+        if (pTree == NULL)
+            return NULL; // no xs:key
+
+        const char* pName = pSchemaRoot->getPropTree(xpath)->queryProp(XML_ATTR_NAME);
+        if (pName != NULL)
+         {
+             pKey = new CKey(pParentNode);
+             pKey->setXSDXPath(xpath);
+             pKey->setName(pName);
+         }
+         else
+         {
+             assert(!"value attribute can be empty!");
+             // TODO: throw MakeExceptionFromMap(EX_STR_MISSING_VALUE_ATTRIBUTE_IN_LENGTH);
+             return NULL;
+         }
+
+         const char *pID = pSchemaRoot->getPropTree(xpath)->queryProp(XML_ATTR_ID);
+         if (pID != NULL)
+             pKey->setID(pID);
+
+         StringBuffer strXPathExt(xpath);
+         strXPathExt.append("/").append(XSD_TAG_FIELD);
+
+         if (strXPathExt.charAt(0) == '@')
+             strXPathExt.remove(0,1); // remove '@'
+
+         pKey->m_pFieldArray = CFieldArray::load(pKey, pSchemaRoot, strXPathExt.str());
+
+         strXPathExt.clear().set(xpath);
+         strXPathExt.append("/").append(XSD_TAG_SELECTOR);
+
+         if (strXPathExt.charAt(0) == '.')
+             strXPathExt.remove(0,2); // remove leading ./
+
+         pKey->m_pSelector = CSelector::load(pKey, pSchemaRoot, strXPathExt.str());
+         assert(pKey->m_pFieldArray != NULL && pKey->m_pSelector != NULL);
+
+         strXPathExt.append("/").append(XSD_TAG_ANNOTATION);
+         pKey->m_pAnnotation = CAnnotation::load(pKey, pSchemaRoot, strXPathExt.str());
+    }
+    return pKey;
+}
+
+void CKey::populateEnvXPath(StringBuffer strXPath, unsigned int index)
+{
+    assert(this->m_pSelector != NULL);
+
+    this->setEnvXPath(strXPath.str());
+
+    if (this->m_pSelector != NULL)
+    {
+        this->m_pSelector->populateEnvXPath(strXPath.str());
+        CConfigSchemaHelper::getInstance()->addKeyForReverseAssociation(this);
+    }
+}
+
+bool CKey::checkConstraint(const char *pValue) const
+{
+    bool bRetVal = true;
+
+    if (m_pSelector != NULL && m_pFieldArray->length() != 0)
+    {
+        for (int idx = 0; idx < m_pFieldArray->length(); idx++)
```

Comment:
The length() can be called before line 112 and use the value in both line 112 and 114.